Implement WebGL GetRenderbufferParameter

fnune opened the pull request:
This needed a bump of gleam to version 0.4.33 for this https://github.com/servo/gleam/pull/162

<!-- Please describe your changes on the following line: -->

I think my changes in test expectations are pretty naive and I have to wait for the PR to run on CI to see what the actual impact is. I'd like some guidance on this, too.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach build-geckolib` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #20514 (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20631)
<!-- Reviewable:end -->

fnune committed:
Implement WebGL GetRenderbufferParameter

This needed a bump of gleam to version 0.4.33

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webglrenderingcontext.rs, components/script/Cargo.toml, components/script/dom/webidls/WebGLRenderingContext.webidl, components/script/dom/webgl2renderingcontext.rs
 * @paulrouget: components/compositing/Cargo.toml, components/servo/Cargo.toml, ports/servo/Cargo.toml
 * @fitzgen: components/script/dom/webglrenderingcontext.rs, components/script/Cargo.toml, components/script/dom/webidls/WebGLRenderingContext.webidl, components/script/dom/webgl2renderingcontext.rs
 * @kichjang: components/script/dom/webglrenderingcontext.rs, components/script/Cargo.toml, components/script/dom/webidls/WebGLRenderingContext.webidl, components/script/dom/webgl2renderingcontext.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

brainlessdeveloper commented:
https://github.com/servo/gleam/pull/162

brainlessdeveloper commented:
r?

gootorov commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -2345,6 +2345,41 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
+        self.send_command(WebGLCommand::GetRenderbufferParameter(target, pname, sender));
+
+        match receiver.recv().unwrap() {
+            value if value != 0 => Int32Value(value),
```
`0` can be a valid result, right?. I think it should be just `Int32Value(receiver.recv().unwrap())`.

gootorov commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 6692241aef655de34d0378f3918411db746317eb with merge a8818d8ae85d7d6fe141a1d4fa83a2110abb111e...

gootorov commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -2345,6 +2345,41 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
 
+    #[allow(unsafe_code)]
+    // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.7
+    unsafe fn GetRenderbufferParameter(&self, _cx: *mut JSContext, target: u32, pname: u32) -> JSVal {
```
For functions with many arguments, we usually indent like so
```Rust
unsafe fn GetRenderbufferParameter(
    &self,
    ...
) -> JSVal {
```

gootorov commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -2345,6 +2345,41 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
+
+        let (sender, receiver) = webgl_channel().unwrap();
+        self.send_command(WebGLCommand::GetRenderbufferParameter(target, pname, sender));
+
```
The following check is missing:
`If  the  renderbuffer  currently  bound  to
target
is  zero,   then
INVALID_-
OPERATION
is generated.`

gootorov replied:
It's just this, basically :)
```Rust
if self.bound_renderbuffer.get().is_none() {
    self.webgl_error(InvalidOperation);
    return NullValue();
}
```

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8500)

brainlessdeveloper commented:
I'll get to this in the evening, thanks :) Also I think my expectations for test failures are wrong, I should be updating more `.ini` files.

bors-servo commented:
:hourglass: Trying commit 6b166ced4e65067394b45cef285be30ab71f025d with merge 6ef4b98d3efbecef14da4f6172aff2c1609b3bb7...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8504)

nox commented:
r? @nox

brainlessdeveloper commented:
I think I just need to update test failure expectations now?

nox commented:
Many more tests now fail in `/_mozilla/webgl/conformance-2.0.0/conformance2/context/methods-2.html` but that's just because the test is running further AFAICT. Update those expectations and we will be able to merge this PR.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20648) made this pull request unmergeable. Please resolve the merge conflicts.

gootorov commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0213688ec117d7ddca9e4f7e52d818444a03408e with merge b035604324c9ca69237122b3f3c6a261173c3a49...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8554)

gootorov commented on tests/wpt/mozilla/meta/webgl/conformance-1.0.3/conformance/context/methods.html.ini:
```diff
@@ -1,7 +1,4 @@
-  [WebGL test #0: Property either does not exist or is not a function: getRenderbufferParameter]
-    expected: FAIL
-
   [WebGL test #1: Property either does not exist or is not a function: getUniform]
```
This index should be updated too.

gootorov commented on components/canvas/webgl_thread.rs:
```diff
@@ -1188,6 +1190,17 @@ impl WebGLImpl {
+        chan: WebGLSender<i32>
+    ) {
+        let parameter = gl.get_renderbuffer_parameter_iv(target, pname);
+        chan.send(parameter).unwrap()
```
nit: Missing semicolon.

brainlessdeveloper commented:
Right, fixed. r?

bors-servo commented:
:hourglass: Trying commit e353bdc332c660d60b398795e24cf1f21f5d9ca9 with merge 2a31d5b2a4982e24b890fd81a629e4052664177b...

brainlessdeveloper commented:
There are more tests failures that my changes don't address:
```
  FAIL WebGL test #57: gl.getUniform(boolProgram, bvalLoc) should be true. Threw exception TypeError: gl.getUniform is not a function                                                                            
  FAIL WebGL test #58: gl.getUniform(boolProgram, bval2Loc) should be true,false. Threw exception TypeError: gl.getUniform is not a function                                                                     
  FAIL WebGL test #59: gl.getUniform(boolProgram, bval3Loc) should be true,false,true. Threw exception TypeError: gl.getUniform is not a function                                                                
  FAIL WebGL test #60: gl.getUniform(boolProgram, bval4Loc) should be true,false,true,false. Threw exception TypeError: gl.getUniform is not a function                                                          
  FAIL WebGL test #63: gl.getUniform(intProgram, ivalLoc) should be 1. Threw exception TypeError: gl.getUniform is not a function                                                                                
  FAIL WebGL test #64: gl.getUniform(intProgram, ival2Loc) should be 2,3. Threw exception TypeError: gl.getUniform is not a function                                                                             
  FAIL WebGL test #65: gl.getUniform(intProgram, ival3Loc) should be 4,5,6. Threw exception TypeError: gl.getUniform is not a function                                                                           
  FAIL WebGL test #66: gl.getUniform(intProgram, ival4Loc) should be 7,8,9,10. Threw exception TypeError: gl.getUniform is not a function                                                                        
  FAIL WebGL test #69: gl.getUniform(floatProgram, fvalLoc) should be 11. Threw exception TypeError: gl.getUniform is not a function                                                                             
  FAIL WebGL test #70: gl.getUniform(floatProgram, fval2Loc) should be 12,13. Threw exception TypeError: gl.getUniform is not a function                                                                         
  FAIL WebGL test #71: gl.getUniform(floatProgram, fval3Loc) should be 14,15,16. Threw exception TypeError: gl.getUniform is not a function                                                                      
  FAIL WebGL test #72: gl.getUniform(floatProgram, fval4Loc) should be 17,18,19,20. Threw exception TypeError: gl.getUniform is not a function                                                                   
  FAIL WebGL test #75: gl.getUniform(samplerProgram, s2DValLoc) should be 0. Threw exception TypeError: gl.getUniform is not a function                                                                          
  FAIL WebGL test #76: gl.getUniform(samplerProgram, sCubeValLoc) should be 1. Threw exception TypeError: gl.getUniform is not a function                                                                        
  FAIL WebGL test #79: gl.getUniform(matProgram, mval2Loc) should be 1,2,3,4. Threw exception TypeError: gl.getUniform is not a function                                                                         
  FAIL WebGL test #80: gl.getUniform(matProgram, mval3Loc) should be 5,6,7,8,9,10,11,12,13. Threw exception TypeError: gl.getUniform is not a function                                                           
  FAIL WebGL test #81: gl.getUniform(matProgram, mval4Loc) should be 14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29. Threw exception TypeError: gl.getUniform is not a function 
```
Are these tests that were legitimately expected to pass?

gootorov commented:
No, `getUniform` is not implemented yet, so it's expected to fail. I guess the failures were raised because of the index thing. It should be fine now.

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8555)

gootorov commented:
Hmm, interesting, `getUniform` isn't set to fail in `gl-object-get-calls.html.ini`, but it now suddenly raises the errors, even though it should have always done that since it isn't implemented, I guess. I thought it raised the errors the previous time because of the index in `methods.ini`. These tests sometimes really confuse me :).

Anyway, updating the expectations in `gl-object-get-calls.html.ini` with that script should resolve the issue.

fnune committed:
Update test failure expectations for gl.getUniform

Updated by running
/mach test-wpt --log-raw /tmp/servo.log ./**/*1.0.3*/conformance/state/gl-object-get-calls.html

...and then
./mach update-wpt /tmp/servo.log

brainlessdeveloper commented:
I placed the updates in a different commit because they have nothing to do with my PR

brainlessdeveloper commented:
r?

gootorov commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 4b4c2865e66568aa00559fdc6a9ed15a14bf6547 with merge d4d800d4f5cbf1161aeaf1f8fe55814f2ccc79d0...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8557)

gootorov commented on tests/wpt/mozilla/meta/webgl/conformance-1.0.3/conformance/state/gl-object-get-calls.html.ini:
```diff
@@ -1,29 +1,53 @@
 [gl-object-get-calls.html]
-  expected: ERROR
-  [WebGL test #33: gl.getRenderbufferParameter(gl.RENDERBUFFER, gl.RENDERBUFFER_WIDTH) should be 2. Threw exception TypeError: gl.getRenderbufferParameter is not a function]
+  expected: TIMEOUT
```
Looking at the last [log](http://build.servo.org/builders/linux-rel-wpt/builds/8557/steps/shell__4/logs/filtered-wpt-errorsummary.log), I guess this should be `OK`, then? Not sure why the script generated this to be a `TIMEOUT` and what's happening with this test in general.

cc @nox to the rescue.

brainlessdeveloper replied:
IDK, I'm pushing again without the `TIMEOUT`

gootorov replied:
Alright, let's see. Hope it works now :).

gootorov commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit ff956827f56d3950f9cc28af0b59505aa831b917 with merge ed6b30fd5cfd9be326b4fd1967f928cbc9e015ce...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11346), [arm32](http://build.servo.org/builders/arm32/builds/11580), [arm64](http://build.servo.org/builders/arm64/builds/11621), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11659), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8552), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8558), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11709), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4104), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4136), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8382), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8443), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4203), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4150), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7051)
State: approved= try=True

brainlessdeveloper commented:
Finally :sweat_smile:

gootorov commented:
> Test successful

cc @nox 
This can be r+'ed

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -2348,6 +2348,45 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
+        if self.bound_renderbuffer.get().is_none() {
+            self.webgl_error(InvalidOperation);
+            return NullValue();
+        }
```
This should probably move after the target and parameter names, to preserve the same order used in https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glGetRenderbufferParameteriv.xml.

emilio reviewed (approved):
Looks good modulo that. Let me know if you prefer to move that in a followup, that's fine for me :)

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11369), [arm32](http://build.servo.org/builders/arm32/builds/11603), [arm64](http://build.servo.org/builders/arm64/builds/11646), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11682), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8578), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8581), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11733), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4127), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4159), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8406), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8468), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4226), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4173), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7074)
Approved by: emilio
Pushing f1a06e0d0c121a2712faf67bc65bd9b297bbf29a to master...
